include/votca/xtp/checkpoint.h

```diff
@@ -0,0 +1,324 @@
+        ReadScalar(_loc, var,  name);
+    }
+private:
+    CptLoc _loc;
+
+    void ReadScalar(const CptLoc& loc, std::string& var, const std::string& name){
+        const H5::DataType* strType = InferDataType<std::string>::get();
+
+        H5::Attribute attr = loc.openAttribute(name);
+
+        H5std_string readbuf("");
+
+        attr.read(*strType, readbuf);
+
+        var = readbuf;
+    }
+
+    template <typename T>
+        void ReadScalar(const CptLoc& loc, T& value,
+                        const std::string& name) {
+
+        H5::Attribute attr = loc.openAttribute(name);
+        const H5::DataType* dataType = InferDataType<T>::get();
+
+        attr.read(*dataType, &value);
+    }
+
+    template <typename T>
+        void ReadData(const CptLoc& loc, Eigen::MatrixBase<T>& matrix,
+                      const std::string& name) {
+
+        const H5::DataType* dataType = InferDataType<typename T::Scalar>::get();
+
+        H5::DataSet dataset = loc.openDataSet(name);
+
+        H5::DataSpace dp = dataset.getSpace();
+
+        hsize_t dims[2];
+        int ndims = dp.getSimpleExtentDims(dims, NULL); // ndims is always 2 for us
+
+        matrix.derived().resize(dims[0], dims[1]);
+        dataset.read(matrix.derived().data(), *dataType);
+    }
+
+    template <typename T>
+        typename std::enable_if<std::is_fundamental<T>::value>::type
+        ReadData(const CptLoc& loc, std::vector<T>& v,
+                 const std::string& name) {
+
+        H5::DataSet dataset = loc.openDataSet(name);
+        H5::DataSpace dp = dataset.getSpace();
+
+        const H5::DataType* dataType = InferDataType<T>::get();
+
+        hsize_t dims[2];
+        int ndims = dp.getSimpleExtentDims(dims, NULL);
+
+        v.resize(dims[0]);
+
+        dataset.read(&(v[0]), *dataType);
+    }
+
+    void ReadData(const CptLoc& loc, votca::tools::vec& v,
+                  const std::string& name){
+
+        // read tools::vec as a vector of three elements
+        std::vector<double> data = {0,0,0};
+        ReadData(loc, data, name);
+        v = votca::tools::vec(data[0], data[1], data[2]);
+    }
+
+    void ReadData(const CptLoc& loc, std::vector<votca::tools::vec>& v,
+                  const std::string& name){
+
+        CptLoc parent = loc.openGroup(name);
+        size_t count = parent.getNumObjs();
+
+        v.resize(count);
+
+        size_t c = 0;
+        for (auto &vec: v){
+            ReadData(parent, vec, "ind"+std::to_string(c));
+            ++c;
+        }
+    }
+};
+
+}  // namespace checkpoint_utils
+
+class CheckpointFile {
+ public:
+  CheckpointFile(std::string fileName, bool overWrite);
+  CheckpointFile(std::string fileName);
+
+  std::string getFileName();
+  std::string getVersion();
+
+  H5::H5File getHandle();
+
+ private:
+  std::string _fileName;
+  H5::H5File _fileHandle;
+  std::string _version;
+};
+
+
+}  // namespace xtp
+}  // namespace votca
+#endif  // CHECKPOINT_H
```

Comment:
Is there any particular reason to include all three classes in the same file? The header file could benefit from some doxygen documentation.

Reply:
Not really, it is mostly laziness on my part. I'll move them to their own files...